File: python/saucebindings/options.py
import warnings
from datetime import datetime

from .configs import *
from selenium.webdriver import __version__ as seleniumVersion
import os


w3c_configs = {
    'browser_name': 'browserName',
    'browser_version': 'browserVersion',
    'platform_name': 'platformName',
    'accept_insecure_certs': 'acceptInsecureCerts',
    'page_load_strategy': 'pageLoadStrategy',
    'proxy': 'proxy',
    'set_window_rect': 'setWindowRect',
    'timeouts': 'timeouts',
    'unhandled_prompt_behavior': 'unhandledPromptBehavior',
    'strict_file_interactability': 'strictFileInteractability'
}

sauce_configs = {
    'avoid_proxy': 'avoidProxy',
    'build': 'build',
    'chromedriver_version': 'chromedriverVersion',
    'command_timeout': 'commandTimeout',
    'custom_data': 'customData',
    'edgedriver_version': 'edgedriverVersion',
    'extended_debugging': 'extendedDebugging',
    'geckodriver_version': 'geckodriverVersion',
    'idle_timeout': 'idleTimeout',
    'iedriver_version': 'iedriverVersion',
    'max_duration': 'maxDuration',
    'name': 'name',
    'parent_tunnel': 'parentTunnel',
    'prerun': 'prerun',
    'priority': 'priority',
    'public': 'public',
    'record_logs': 'recordLogs',
    'record_screenshots': 'recordScreenshots',
    'record_video': 'recordVideo',
    'screen_resolution': 'screenResolution',
    'selenium_version': 'seleniumVersion',
    'tags': 'tags',
    'time_zone': 'timeZone',
    'tunnel_identifier': 'tunnelIdentifier',
    'video_upload_on_pass': 'videoUploadOnPass',
    'capture_performance': 'capturePerformance'
}

class SauceOptions(object):

    @classmethod
    def chrome(cls, **kwargs):
        return cls('chrome', validOptions=Configs().chromeConfigs(), **kwargs)

    @classmethod
    def edge(cls, **kwargs):
        if seleniumVersion[0] == '3':
            raise NotImplementedError('Selenium 3 does not support Chromium Edge. Look for SauceBindings Support of '
                                      'Selenium 4 soon.')
        return cls('MicrosoftEdge', validOptions=Configs().edgeConfigs(), **kwargs)

    @classmethod
    def firefox(cls, **kwargs):
        return cls('firefox', validOptions=Configs().firefoxConfigs(), **kwargs)

    @classmethod
    def ie(cls, **kwargs):
        return cls('internet explorer', validOptions=Configs().ieConfigs(), **kwargs)

    @classmethod
    def safari(cls, **kwargs):
        if 'platformName' not in kwargs:
            kwargs['platformName'] = 'macOS 11.00'

        return cls('safari', validOptions=Configs().safariConfigs(), **kwargs)

    def _set_default_build_name(self):
        if 'build' in self.options['sauce:options']:
            return
        # Jenkins
        elif os.environ.get("BUILD_TAG"):
            self.build = "{}: {}".format(os.environ['BUILD_NAME'], os.environ['BUILD_NUMBER'])
        # Bamboo
        elif os.environ.get('bamboo_agentId'):
            self.build = "{}: {}".format(os.environ['bamboo_shortJobName'], os.environ['bamboo_buildNumber'])
        # Travis
        elif os.environ.get('TRAVIS_JOB_ID'):
            self.build = "{}: {}".format(os.environ['TRAVIS_JOB_NAME'], os.environ['TRAVIS_JOB_NUMBER'])
        # Circle
        elif os.environ.get('CIRCLE_JOB'):
            self.build = "{}: {}".format(os.environ['CIRCLE_JOB'], os.environ['CIRCLE_BUILD_NUM'])
        # GitHub Actions
        elif os.environ.get("GITHUB_SHA"):
            self.build = "{}: {}".format(os.environ['GITHUB_WORKFLOW'], os.environ['GITHUB_SHA'])
        # Gitlab
        elif os.environ.get('CI_JOB_ID'):
            self.build = "{}: {}".format(os.environ['CI_JOB_NAME'], os.environ['CI_JOB_ID'])
        # Team City
        elif os.environ.get('TEAMCITY_PROJECT_NAME'):
            self.build = "{}: {}".format(os.environ['TEAMCITY_PROJECT_NAME'], os.environ['BUILD_NUMBER'])
        else:
            self.build = 'Build Time: {}'.format(datetime.utcnow())

    def __init__(self, browserName=None, validOptions=None, seleniumOptions=None, **kwargs):
        super(SauceOptions, self).__setattr__('options', {'sauce:options': {}})
        super(SauceOptions, self).__setattr__('seleniumOpts', {})
        if validOptions == None:
            warnings.warn('Options() is deprecated, use class methods like Options.chrome() instead',
                          DeprecationWarning)
            validOptions = {**w3c_configs, **sauce_configs}
        super(SauceOptions, self).__setattr__('validOptions', validOptions)

        if self.validOptions == None:
            self.validOptions = {**w3c_configs, **sauce_configs}

        self.validateOptions(kwargs)

        if seleniumOptions is not None:
            self.set_capability('browserName', seleniumOptions.capabilities['browserName'])
            self.seleniumOpts['caps'] = seleniumOptions.to_capabilities()

        for key, value in kwargs.items():
            self.set_capability(key, value)

        if self.browser_version is None:
            self.set_capability('browserVersion', 'latest')

        if browserName:
            self.set_capability('browserName', browserName)
        elif self.browser_name:
            pass  # browser name set by Selenium Class
        else:
            self.set_capability('browserName', 'chrome')

        if self.platform_name is None:
            self.set_capability('platformName', 'Windows 10')

        self._set_default_build_name()

    def __setattr__(self, key, value):
        if (self.validOptions == None and key == 'validOptions') or key in self.validOptions.keys():
            self.set_option(key, value)
        else:
            raise AttributeError('parameter ' + key + ' not available for this configuration')

    def __getattr__(self, key):
        if key == 'selenium_options':
            if 'caps' in self.seleniumOpts.keys():
                return self.seleniumOpts['caps']
            else:
                return None
        try:
            if key in sauce_configs.keys():
                return self.options['sauce:options'][sauce_configs[key]]
            elif key in w3c_configs.keys():
                return self.options[w3c_configs[key]]
            else:
                raise AttributeError
        except KeyError:
            return None

    def validateOptions(self, kwargs):
        for k in kwargs.keys():
            if k not in self.validOptions.values():
                raise AttributeError('parameter ' + k + ' not available for this configuration')

    def merge_capabilities(self, capabilities):
        for key, value in capabilities.items():
            if key not in self.validOptions.values():
                raise AttributeError('parameter ' + key + ' not available for this configuration')
            self.set_capability(key, value)

    # Sets with camelCase
    def set_capability(self, key, value):
        if key in sauce_configs.values():
            self.options['sauce:options'][key] = value
        elif key in w3c_configs.values():
            self.options[key] = value
        else:
            raise AttributeError

    # Sets with snake_case
    def set_option(self, key, value):
        if key in sauce_configs.keys():
            self.options['sauce:options'][sauce_configs[key]] = value
        elif key in w3c_configs.keys():
            self.options[w3c_configs[key]] = value
        else:
            raise AttributeError

    def to_capabilities(self):
        if self.selenium_options:
            self.options.update(self.selenium_options)
        return self.options

File: ruby/lib/sauce_bindings/options.rb
# frozen_string_literal: true

require 'selenium-webdriver'

module SauceBindings
  class Options
    class << self
      def camel_case(str)
        str.to_s.gsub(/_([a-z])/) { Regexp.last_match(1).upcase }
      end

      def chrome(**opts)
        opts[:browser_name] = 'chrome'
        opts[:valid_options] = ChromeConfigurations.valid_options
        Options.new(**opts)
      end

      def edge(**opts)
        if Selenium::WebDriver::VERSION[0] == '3'
          raise ArgumentError, 'Selenium 3 is not compatible with the Chromium based Microsoft Edge.'
        end

        opts[:browser_name] = 'MicrosoftEdge'
        opts[:valid_options] = EdgeConfigurations.valid_options
        Options.new(**opts)
      end

      def firefox(**opts)
        opts[:browser_name] = 'firefox'
        opts[:valid_options] = FirefoxConfigurations.valid_options
        Options.new(**opts)
      end

      def ie(**opts)
        opts[:browser_name] = 'internet explorer'
        opts[:valid_options] = IEConfigurations.valid_options
        Options.new(**opts)
      end

      def safari(**opts)
        opts[:browser_name] = 'safari'
        opts[:platform_name] ||= 'macOS 11'
        opts[:valid_options] = SafariConfigurations.valid_options
        Options.new(**opts)
      end
    end

    BROWSER_NAMES = {Selenium::WebDriver::Chrome::Options => 'chrome',
                     Selenium::WebDriver::Edge::Options => 'MicrosoftEdge',
                     Selenium::WebDriver::Firefox::Options => 'firefox',
                     Selenium::WebDriver::IE::Options => 'internet explorer',
                     Selenium::WebDriver::Safari::Options => 'safari'}.freeze

    W3C = %i[browser_name browser_version platform_name accept_insecure_certs page_load_strategy proxy
             set_window_rect timeouts unhandled_prompt_behavior strict_file_interactability].freeze

    SAUCE = %i[avoid_proxy build chromedriver_version command_timeout custom_data extended_debugging idle_timeout
               iedriver_version max_duration name parent_tunnel prerun priority public record_logs record_screenshots
               record_video screen_resolution selenium_version tags time_zone tunnel_identifier video_upload_on_pass
               capture_performance].freeze

    attr_reader :selenium_options, :timeouts

    def initialize(**opts)
      se_options = Array(opts.delete(:selenium_options))
      valid_options = opts.delete(:valid_options)
      if valid_options.nil?
        valid_options = SAUCE + W3C
        warn 'Using `Options.new(opts)` directly is deprecated, use static methods for desired browser instead; ' \
        'e.g., `Options.chrome(opts)` or `Options.safari(opts)'
      end

      create_variables(valid_options, opts)
      @selenium_options = se_options.map(&:as_json).inject(:merge) || {}
      parse_selenium_options(se_options)
      @build ||= build_name

      @browser_name ||= selenium_options['browserName'] || 'chrome'
      @platform_name ||= 'Windows 10'
      @browser_version ||= 'latest'
    end

    def capabilities
      caps = selenium_options.dup
      W3C.each do |key|
        value = parse_w3c_key(key)
        caps[self.class.camel_case(key)] = value unless value.nil?
      end
      caps['sauce:options'] = {}
      SAUCE.each do |key|
        value = parse_sauce_key(key)
        caps['sauce:options'][self.class.camel_case(key)] = value unless value.nil?
      end
      caps
    end
    alias as_json capabilities

    def merge_capabilities(opts)
      opts.each do |key, value|
        raise ArgumentError, "#{key} is not a valid parameter for Options class" unless respond_to?("#{key}=")

        value = value.transform_keys(&:to_sym) if value.is_a?(Hash)

        send("#{key}=", value)
      end
    end

    private

    def key_value(key)
      send(key)
    rescue NoMethodError
      # Ignored
    end

    def parse_w3c_key(key)
      value = send(key)

      if key == :proxy && value
        value.as_json
      elsif key == :timeouts
        if value
          warn ':timeouts is deprecated, use :implicit_wait_timeout, :page_load_timeout or :script_timeout directly; ' \
          '(ensure the values are set as seconds not milliseconds)'
          value.transform_keys do |old_key|
            self.class.camel_case(old_key)
          end
        else
          manage_timeouts
        end
      else
        value
      end
    end

    def manage_timeouts
      timeouts = {}
      timeouts['implicit'] = ms_multiplier(@implicit_wait_timeout) if @implicit_wait_timeout
      timeouts['pageLoad'] = ms_multiplier(@page_load_timeout) if @page_load_timeout
      timeouts['script'] = ms_multiplier(@script_timeout) if @script_timeout
      timeouts unless timeouts.empty?
    end

    def ms_multiplier(int)
      raise(ArgumentError, 'Timeouts need to be entered as seconds not milliseconds') if int > 600

      int * 1000
    end

    def parse_sauce_key(key)
      value = key_value(key)

      case key
      when :prerun
        camelize_keys(value)
      when :custom_data
        camelize_keys(value)
      when :disable_record_video
        !send(:record_video)
      when :disable_video_upload_on_pass
        !send(:video_upload_on_pass)
      when :disable_record_screenshots
        !send(:record_screenshots)
      when :disable_record_logs
        !send(:record_logs)
      else
        value
      end
    end

    def camelize_keys(hash)
      hash&.transform_keys do |old_key|
        self.class.camel_case(old_key)
      end
    end

    def parse_selenium_options(selenium_opts)
      selenium_opts.each do |opt|
        W3C.each do |capability|
          value = option_value(opt, capability)
          next if capability == :timeouts && value&.empty?

          if capability == :browser_name
            @browser_name ||= BROWSER_NAMES[opt.class]
            validate_browser_name(value || BROWSER_NAMES[opt.class])
          elsif value
            send("#{capability}=", value)
          end
        end
      end
    end

    def option_value(opt, capability)
      if opt.respond_to?(capability)
        begin
          opt.send(capability)
        rescue KeyError
          # Ignored
        end
      elsif opt.respond_to?(:[])
        opt[capability]
      end
    end

    def validate_browser_name(browser)
      return if @browser_name.nil? || browser == @browser_name

      raise ArgumentError, "Selenium class identifies capabilities for #{browser}, which does not match the " \
"provided browser name: #{@browser_name}"
    end

    def create_variables(key, opts)
      key.each do |option|
        singleton_class.class_eval { attr_accessor option }
        next unless opts.key?(option)

        instance_variable_set("@#{option}", opts.delete(option))
      end

      if opts.key?(:browser_name)
        singleton_class.class_eval { attr_reader :browser_name }
        @browser_name = opts.delete(:browser_name)
      end

      return if opts.empty?

      raise ArgumentError, "#{opts.inspect} are not valid parameters for Options class with #{@browser_name}"
    end

    def build_name
      # Jenkins
      if ENV['BUILD_TAG']
        "#{ENV['BUILD_NAME']}: #{ENV['BUILD_NUMBER']}"
      # Bamboo
      elsif ENV['bamboo_agentId']
        "#{ENV['bamboo_shortJobName']}: #{ENV['bamboo_buildNumber']}"
      # Travis
      elsif ENV['TRAVIS_JOB_ID']
        "#{ENV['TRAVIS_JOB_NAME']}: #{ENV['TRAVIS_JOB_NUMBER']}"
      # CircleCI
      elsif ENV['CIRCLE_JOB']
        "#{ENV['CIRCLE_JOB']}: #{ENV['CIRCLE_BUILD_NUM']}"
      # GitHub Actions
      elsif ENV['GITHUB_SHA']
        "#{ENV['GITHUB_WORKFLOW']}: #{ENV['GITHUB_SHA']}"
      # Gitlab
      elsif ENV['CI_JOB_NAME']
        "#{ENV['CI_JOB_NAME']}: #{ENV['CI_JOB_ID']}"
      # Team City
      elsif ENV['TEAMCITY_PROJECT_NAME']
        "#{ENV['TEAMCITY_PROJECT_NAME']}: #{ENV['BUILD_NUMBER']}"
      # Default
      else
        "Build Time - #{Time.now.to_i}"
      end
    end
  end
end
